WSLBlobNFS/WSLBlobNFS.psm1

```diff
@@ -6,178 +11,539 @@ Set-PSDebug -Strict
+    if($Force)
+    {
+        Write-Warning "Force initializing WSL environment for WSLBlobNFS usage."
+    }
+    else
+    {
+        Write-Debug "Initializing WSL environment for your WSLBlobNFS usage."
+    }
 
-    Write-Host "Installing Ubuntu-22.04..."
-    wsl --install -d Ubuntu-22.04
+    # # Add the modulePathForLinux, which hold the path of module, to WSLENV to help in copying linux scripts to wsl distro.
+    # [Environment]::SetEnvironmentVariable("modulePathForLinux", $PSScriptRoot)
+    # $p = [Environment]::GetEnvironmentVariable("WSLENV")
+    # $p += ":modulePathForLinux/p"
+    # [Environment]::SetEnvironmentVariable("WSLENV",$p)
 
-    # ask user if the username was setup
-    $usernameSetup = Read-Host "Was the wsl distro username setup? (y/n)"
-    if($usernameSetup -ne "y")
+    $moduleVersionOnWin = Get-ModuleVersion
+    Write-Debug "Module version on Windows: $moduleVersionOnWin"
+    $moduleVersionOnWSL = Execute-WSL "cat $linuxScriptsPath/$versionFileName 2>/dev/null"
+    Write-Debug "Module version on WSL: $moduleVersionOnWSL"
+    if(($LASTEXITCODE -eq 0) -and ($moduleVersionOnWSL.Trim() -eq $moduleVersionOnWin) -and !$Force)
     {
-        Write-Host "Restart the VM to complete wsl installation, and then run the script again with setupwslenv action to continue the wsl setup."
-        return
+        Write-Debug "WSL Files are upto date. Skipping copying files."
+    }
+    else
+    {
+        $initialized = $true
+        if($Force)
+        {
+            Write-Debug "Force parameter is provided. Copying files again."
+        }
+        elseif ($LASTEXITCODE -ne 0)
+        {
+            Write-Debug "Module is not installed. Copying files again."
+        }
+        else
+        {
+            Write-Debug "Module version is changed. Copying files again."
+        }
+
+        Copy-WSLFiles
+        if ($LASTEXITCODE -ne 0)
+        {
+            Write-Error "Copying wsl files failed."
+            return
+        }
+
+        Execute-WSL "echo $moduleVersionOnWin > $linuxScriptsPath/$versionFileName"
     }
 
-    Write-Host "Setting up the WSL environment for your WSLBlobNFS usage."
+    # Install systemd, restart wsl and install NFS & Samba
+    Execute-WSL "systemctl list-unit-files --type=service | grep systemd > /dev/null 2>&1"
+    if($LASTEXITCODE -eq 0 -and !$Force)
+    {
+        Write-Debug "Systemd is already installed. Skipping systemd installation."
+    }
+    else
+    {
+        $initialized = $true
+        if($Force)
+        {
+            Write-Debug "Force parameter is provided. Installing systemd again."
+        }
+        else
+        {
+            Write-Debug "Systemd is not installed. Installing systemd."
+        }
+
+        Execute-WSL "$linuxScriptsPath/$wslScriptName installsystemd"
+        Write-Verbose "Installed systemd."
+
+        # Shutdown wsl and it will restart with systemd on next wsl command execution
+        wsl -d $distroName --shutdown
```

Comment:
there should be a log telling this, since restarting wsl is fairly intrusive, as user may be running something in wsl while he is installing this
Infact we should also ask user's permission to do that